src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2836,8 +2834,31 @@ module Crystal
       false
     end
 
+    def visit(node : TypeRestrict)
+      value = node.value
+      restriction = node.restriction
+
+      value.accept self
+
+      unless context = match_context
```

Comment:
I'm thinking, maybe there's a different way to do this. The problem with this approach is that the type is checked once here (below). But what happens if the type of `value` changes later? (rare, but can happen)

Another way to do this is to do something like:

```cr
# This line is currently in `visit(node : Path)`, so we could probably extract it to a helper method
restriction_type = (@path_lookup || @scope || @current_type).lookup_type_var(node, free_vars: @free_vars)
unless restriction_type.is_a?(Type)
  node.raise "restriction must be a type, not #{restriction_type}"
end

# With this we make the node always have this type, and raise an error if it's not (also after a change)
value.freeze_type = restriction_type
```

In this way I think there's no need to change the visitor to have a `MatchContext` instead of `free_vars`.

Could you try this change and see if it works?